pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,407 @@
+		value := br.ReadVarBytes()
+		if br.Err != nil {
+			bfs.log.Error("failed to read value", zap.Error(br.Err))
+			return br.Err
+		}
+
+		fullKey := make([]byte, len(tempPrefix)+len(neoFSKey))
+		copy(fullKey, tempPrefix)
+		copy(fullKey[len(tempPrefix):], neoFSKey)
+		key := neoFSKey
+		if processed < 5 {
+			bfs.log.Debug("key sample", zap.ByteString("fullKey", fullKey), zap.ByteString("mptKey", key), zap.ByteString("value", value))
+		}
+
+		cache.Put(fullKey, value)
+		batch[string(key)] = value
+		processed++
+
+		if len(batch) >= batchSize {
+			if err := bfs.processMPTBatch(localMPT, batch, syncHeight, cache); err != nil {
+				bfs.log.Error("failed to process MPT batch", zap.Error(err))
+				return err
+			}
+			if n, err := cache.PersistSync(); err != nil {
+				bfs.log.Error("failed to persist cache", zap.Error(err), zap.Int("items", n))
+				return err
+			}
+			bfs.log.Info("batch root", zap.String("root", localMPT.StateRoot().String()))
+			batch = make(map[string][]byte, batchSize)
+			cache = storage.NewMemCachedStore(bfs.stateModule.Store)
+			bfs.log.Info("processed and persisted MPT batch", zap.Int("items", processed))
+		}
+	}
+
+	if len(batch) > 0 {
+		if err := bfs.processMPTBatch(localMPT, batch, syncHeight, cache); err != nil {
+			bfs.log.Error("failed to process final MPT batch", zap.Error(err))
+			return err
+		}
+		if n, err := cache.PersistSync(); err != nil {
+			bfs.log.Error("failed to persist final cache", zap.Error(err), zap.Int("items", n))
+			return err
+		}
+	}
+
+	computedRoot := localMPT.StateRoot()
+	bfs.log.Info("computed MPT root", zap.String("root", computedRoot.String()))
+
+	if !computedRoot.Equals(stateRoot) {
+		bfs.log.Error("state root mismatch",
+			zap.String("computed", computedRoot.String()),
+			zap.String("neofs", stateRoot.String()))
+		return fmt.Errorf("state root mismatch: expected %s, got %s", stateRoot, computedRoot)
+	}
+	if err := bfs.finalizeState(syncHeight, cache); err != nil {
+		bfs.log.Error("failed to finalize state", zap.Error(err))
+		return err
+	}
+
+	bfs.log.Info("completed MPT state sync",
+		zap.Uint32("height", syncHeight),
+		zap.Int("total_items", processed),
+		zap.String("root", computedRoot.String()))
+	return nil
+}
+
+func (bfs *Service) processMPTBatch(trie *mpt.Trie, changes map[string][]byte, height uint32, cache *storage.MemCachedStore) error {
+	batch := mpt.MapToMPTBatch(changes)
+	trie.Store = cache
+	if _, err := trie.PutBatch(batch); err != nil {
+		return fmt.Errorf("failed to apply MPT batch at height %d: %w", height, err)
+	}
+	trie.Flush(height)
+	bfs.log.Info("applied MPT batch", zap.Int("items", len(changes)), zap.String("new_root", trie.StateRoot().String()))
+	return nil
+}
+func (bfs *Service) getMPTNodes() (io.ReadCloser, error) {
+	headerHeight := bfs.chain.HeaderHeight() - 1
+	var stateOidsObject []oid.ID
+	var err error
+	ctx, cancel := context.WithTimeout(bfs.ctx, bfs.cfg.Timeout)
+	defer cancel()
+	for i := headerHeight; true; i-- {
```

Comment:
We have already discussed that, we need to search for the nearest state sync point, not every block for sure.